clippy_lints/src/methods/option_map_unwrap_or.rs

```diff
@@ -56,24 +59,19 @@ pub(super) fn lint<'a, 'tcx>(
              This can be done more directly by calling `{}` instead",
             arg, suggest
         );
-        // lint, with note if neither arg is > 1 line and both map() and
-        // unwrap_or() have the same span
-        let multiline = map_snippet.lines().count() > 1 || unwrap_snippet.lines().count() > 1;
-        let same_span = map_args[1].span.ctxt() == unwrap_args[1].span.ctxt();
-        if same_span && !multiline {
-            let suggest = if unwrap_snippet == "None" {
-                format!("and_then({})", map_snippet)
-            } else {
-                format!("map_or({}, {})", unwrap_snippet, map_snippet)
-            };
-            let note = format!(
-                "replace `map({}).unwrap_or({})` with `{}`",
-                map_snippet, unwrap_snippet, suggest
-            );
-            span_note_and_lint(cx, OPTION_MAP_UNWRAP_OR, expr.span, msg, expr.span, &note);
-        } else if same_span && multiline {
-            span_lint(cx, OPTION_MAP_UNWRAP_OR, expr.span, msg);
-        };
+
+        span_lint_and_then(cx, OPTION_MAP_UNWRAP_OR, expr.span, msg, |db| {
+                db.multipart_suggestion(
+                    "use this instead",
+                    vec![
+                        (map_span, String::from("map_or")),
+                        (expr.span.with_lo(unwrap_args[0].span.hi()), String::from("")),
+                        (map_arg_span.with_hi(map_arg_span.lo()), format!("{}, ", unwrap_snippet))
+                    ],
+                    Applicability::MaybeIncorrect
```

Comment:
This could probably even be `Applicability::MachineApplicable` modulo the caveat outlined above.